### ISSUE TITLE
[Backend] - Open-Ended Questions Adjustment Proposal

### ISSUE THREAD

**Issue by drewjhart** [rightoneducation/righton-app#789]
# Issue: Backend Adjustment  #

## Overview ##
We have been working with the Design team to implement some new features in the past few weeks. The Figma is [here](https://www.figma.com/file/PadzU8GqyhFLCjIt7rQkpE/RightOn-Apps?type=design&node-id=15528-176696&mode=dev) and this is a breakdown:

1. Open - Ended Questions - provides a text box for the students to enter their answer in, instead of multiple-choice:

Host:
<img width="317" alt="Screenshot 2023-08-09 at 9 21 49 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/052c2077-010c-41fd-b961-16a8ed09b258">

Play:
<img width="206" alt="Screenshot 2023-08-09 at 9 21 30 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/b95897c7-8695-477a-879e-5dd07cc53934">

2. Confidence Meter - tracks a student's confidence in their phase one answer

Host:
<img width="317" alt="Screenshot 2023-08-09 at 9 21 53 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/9792b111-48aa-425f-8a1f-f29e161b81ea">

Play:
<img width="210" alt="Screenshot 2023-08-09 at 9 21 17 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/3fada4dd-b37a-4e02-b1fc-eafa61e9f585">

3. Student Thinking - allows a student to enter a hint for their phase two answer, providing some insight into their thinking

Host:
<img width="207" alt="Screenshot 2023-08-09 at 9 21 08 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/9315dba6-fbeb-4f08-bb85-9ea9a47f4a8b">

Play:
<img width="211" alt="Screenshot 2023-08-09 at 9 20 55 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/a450bdee-2250-4635-929b-6394c0392d1a">

### Game Flow Changes ###
As a result of these features, the Design team has proposed the following changes: 

1. Open - Ended Questions and the Confidence Meter will be enabled and disabled in a new `GameSessionState` that occurs before the start of each question. Essentially we will move from `TEAMS_JOINING` to the new phase and then to `CHOOSE_CORRECT_ANSWER.` I am calling this phase `PREQUESTION`. This phase will contain the following controls:

<img width="224" alt="Screenshot 2023-08-09 at 9 16 42 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/6cd9d323-3bda-4cb2-901e-c074cedf4b2a">

Enabling the teacher to set what kind of question and feedback the students will see before beginning each question. At the end of each question `PHASE_2_DISCUSS`, we will return back to `PREQUESTION`. On the `play` side, the student will see a leaderboard of the current scores and a loading indicator.

2. Student Thinking will be enabled during `CHOOSE_CORRECT_ANSWER`, allowing the teacher to respond to the type of answers that they are seeing in the Featured Mistakes component: 

<img width="431" alt="Screenshot 2023-08-09 at 9 19 16 AM" src="https://github.com/rightoneducation/righton-app/assets/79417944/f4bb3607-dd6a-4522-b5d6-569e8001252c">

If the switch in the top right of the component is selected, students will see the "Provide a hint dialogue" 

## Proposed Adjustments ##
With the above in mind, I am proposing the following changes to our `schema`:

1. GameSesssionState changes:
```
enum GameSessionState {
  NOT_STARTED
  TEAMS_JOINING
  PREQUESTION // <---- NEW
  CHOOSE_CORRECT_ANSWER
  PHASE_1_DISCUSS
  PHASE_1_RESULTS
  PHASE_2_START
  CHOOSE_TRICKIEST_ANSWER
  PHASE_2_DISCUSS
  PHASE_2_RESULTS
  FINAL_RESULTS
  FINISHED
}
```
Required as a new state in which the teacher can adjust controls for the question type.

2. Question changes:
```
type Question @model(subscriptions: null, timestamps: null) {
  id: Int! @primaryKey(sortKeyFields: ["order", "gameSessionId"])
  text: String!
  choices: AWSJSON
  imageUrl: String
  instructions: AWSJSON
  standard: String
  cluster: String
  domain: String
  grade: String
  order: Int!
  questionType: QuestionType! // <---- NEW
  enableConfidence: Boolean! // <---- NEW
  enableHint: Boolean! // <---- NEW
  gameSessionId: ID! @index(name: "byGameSession", sortKeyFields: ["id"])
}

enum QuestionType { // <---- NEW
  MULTIPLE_CHOICE
  OPEN_ENDED
}
```

Adds three new fields to the question object, one that tracks the type of question the teacher has selected, one that tracks if the confidence meter will be enabled for students and one that will track if the hint feature is enabled for students. These will be manipulated during the `PREQUESTION` and `CHOOSE_CORRECT_ANSWER` phases on the `host` side via the `UpdateQuestion` mutation, using the `id` of the question and the `gameSessionId` of the current gameSession.

## Implications ##
I think this will be a similar change to the confidenceLevel change that Alex implemented a few weeks ago. It will invalidate questions on the backend (as they dont have the three new required fields) but it shouldn't directly break anything, as we aren't changing any existing fields/types. I will update the backend similarly to how Alex did, first apply these changes via `amplify` pushing to the backend and releasing a PR. I will then adjust the `networking` layer so that the parsing functions surrounding the `question` object capture the new fields.

**Comment by maniramezan** [comment]
hmm, just on top of my head, here's some thoughts:

- Central currently doesn't force the user to enter multiple choices. So during game creation, we do validations on backend API to make sure all questions have multiple choices for basic game before letting you to create one. Why don't we follow the same pattern? When a teacher is making a game session, we show the modal to choose the options for multiple choice / open and confidence level. Then our backend API for creating game session can handle validations. This way, we won't have any changes to flow of the data
- We initially talked about a question can be used for both open-ended and multiple choices, that's why we added `choices` property to a question. If `choices` has only one field, that means the question can only be used in open-ended. If the question has multiple choices (currently hardcoded at 4), this question can be used for both open-ended and multiple choices. This design gives us the flexibility of re-using the same question for both purposes. 
- If we want to give control per question, meaning have a game session with mixture of questions and settings, that's still possible to be done during game session creation instead of game itself.

Your proposed changes for `Question` model looks fine to me. If we can decide on game session creation what to do with questions, that's probably better and we can omit adding a new state. The only benefit with adding a new state is letting teachers to decide what to do with each questions during game session depending on the feedback they receive from students

**Comment by drewjhart** [comment]
Thanks for the response on this. I'm going to loop @amarax into this as well because we were talking yesterday about the kind of control we're looking to give the teacher during the game. I raised the idea of just adjusting these settings once at the start of the game but I think Yong Lin, you're looking for this to be kind of an on-the-fly functionality, right? I think the idea is that the teacher can adjust the style of the questions to accommodate for the classes performance/engagement in previous questions. Correct me if I'm wrong on that though, Yong Lin.

Also, there is the desire to show the leaderboard at the start or end of every question. Conceivably, we could just pop up a modal on `PHASE_2_DISCUSS`  or 'CHOOSE_CORRECT_ANSWER'  phases, but then we would need to sequence that modal so that it didn't interrupt the existing state (kids didn't get distracted by the scores and not pay attention to wrong answer explanations, for instance). I feel like if we already have this pause for the teacher to manipulate the scores, it would be a good moment for the students to review the scoreboard without interrupting their engagement during answering a question or reviewing those answers

**Comment by amarax** [comment]
I would say that the on-the-fly functionality is ideal, because of the overall concept of having a "lesson plan", where the teacher may start with simpler or warm up questions (likely multiple choice), and then progress to the learning objective of the lesson (likely involving open ended).

In an ideal world, this would look like a game (on Central) having a recommended sequence and question settings, but eventually allowing teachers to override them to match the needs of the students. 

We should note that the formulation of the question itself very likely will be affected by how it is answered, i.e. an open-ended question probably is designed to be answered in an open-ended way.

To summarise and clarify:
1. We will want game sessions to have a mix of multiple-choice and open-ended questions
2. In the future, we should work towards Central having indications of what type of questions there should be
3. We will likely want to support teachers to have their own overrides of the question type during game creation and during the game session itself

I would like to hear what would be recommended to quickly get to 1, while considering that 2 and 3 are on the roadmap. Also looping @sincwu to confirm that this makes sense when considering the overall product direction!

**Comment by sincwu** [comment]
Nice summaries, Drew and Yong Lin! Agree with all that was discussed yesterday and recapped above.

**Comment by maniramezan** [comment]
Thanks @amarax and @sincwu for confirmation. In that case, I think we should spend a bit more time on scoping the work and think about phase releasing this. Here're my questions:

- Is this a new game mode or a new way of answering questions only. This can be used with different game modes
- @drewjhart Something to call out is to see which new fields would require formula changes as well. Like, are we going to add formula into hints or not? This affects the model changes

Here're my suggestion:

- If we're ok with central defining how each question should be used, then a lot of complexity for validating questions and make sure everything is ready before starting the game can be done in central. This can be the phase one
- Adding the ability to modify the questions on-the-fly when the game session is in-progress can be built upon the previous work. This would require extra validation works on `Host` to make sure a question for example can be used in a multiple-choice mode. This work can also be done in Central by enforcing all questions are eligible for both scenarios

We should also think about what's the relationship between this feature vs. our current play modes. Is this replacing those or used on the side with each? If latter, what does it mean for each mode?

**Comment by drewjhart** [comment]
These are great follow ups, thanks Mani. As I read through this discussion, I'm wondering if people have time next week (I'm off this Friday) to talk through this in more detail. I personally feel that there's a lot of thought put into the backend that I'm not currently caught up on and I know Yong Lin has also got a ton of thinking on the new Intermediate mode. It might be productive to just get online and talk through the various threads here. 

Given that we'd be spanning EST to SST timezones, is there a spot that works for anyone next week?  Mani, I know we usually do afternoons but that is the middle of the night for Yong Lin. Do you have anything in the morning by any chance?  We've done a 10am EST that worked ok  (if that's still good for you, Yong Lin).

**Comment by maniramezan** [comment]
Sounds good, yeah 10am works for me almost any day. Free 10am-11am, other than not being a morning person, my mornings before 11am is almost free every day if needed

**Comment by drewjhart** [comment]
hahaha I'll need some coffee too. That's great though, thanks for finding some availability. @amarax and @sincwu is there a day next week at 10am EST that works ok for you both?

**Comment by sincwu** [comment]
Really good points, @maniramezan, about how, if all questions can be played in either multiple-choice or open-ended formats, that'd mean all questions would have to have predefined distractors and wrong answer explanations associated with them. This we know is one of the biggest hurdles to game/question setup. We'll be meeting with a couple teachers over the next few days, and we'll try to get some additional feedback and perspectives on this. 

Over time, we'll likely want to further segment our target user personas: e.g., newer / younger teachers who may want a simpler app that they can run on auto-pilot and more seasoned educators who actively work to create a positive culture of error in their classrooms and have implemented / would like to implement error-based teaching methods. The latter is a much smaller group, but their support of and interest in RightOn is also greater.

A team meeting to flesh out plans further sounds good! Yong Lin, Drew, and I have a regular sync next Tuesday, 8/15 at 10am ET (on our team calendar); am thinking we might use that time to continue the conversation, first focusing on product requirements, considering the points brought up in this thread and any additional teacher feedback we'll have been able to gather by then. 

Hi Mani, To be mindful of your time: while you're of course welcome to join that 8/15 10am ET meeting to get a preview into our plans and share technical considerations that may come to mind, I'm thinking that a 2nd meeting, one focused on technical requirements and plans would be helpful. That 2nd meeting could perhaps be Thursday, 8/17 at 10am ET (if that time also works for Yong Lin) or the following Tuesday, 8/22 at 10am ET. The above are all just suggestions -- definitely open to others -- which would you prefer?

**Comment by maniramezan** [comment]
Thanks @sincwu! Both those dates work for me.

**Comment by sincwu** [comment]
Great, looking forward to chatting with you and everyone tomorrow! A few quick updates:
1. Met with 3 math teachers over the past few days, including a former math educator / instructional coach with 10+ years of experience: please find notes [here](https://docs.google.com/document/d/14mcOlJw0i6YfdJqonbGNkpMbuqyWBdIGwFjqF3khW5c/edit?usp=sharing); tldr: open-ended questions reveal more student thinking than multiple-choice questions, but they also take more time; err on the side of simplicity when designing features
2. No one's got a crystal ball, so we're going to just continue to do our best with the info we have and plan to iterate as we learn more: based on the latest feedback, am inclined to continue to lean toward designing/developing for newer/younger teachers who might be more tech-savvy but have less math pedagogy experience -- this is a much larger group of prospective users, and they have the most to gain
3. +1 to the suggestion/plan summarized by Drew above where 3 new fields will be added to the question object (will defer technical implementation details to Drew/Mani; the functional spec is on point) 
> _one that tracks the type of question the teacher has selected, one that tracks if the confidence meter will be enabled for students and one that will track if the hint feature is enabled for students_
4. For our upcoming team sync (Tue 10am EST / Tue 10pm SST): our going-in "strawman" (feel free to tear things apart :)) might be: a) default all questions to multiple-choice; confidence meter: disabled; prompt for students to enter an optional hint: enabled, b) provide an on-the-fly option on the Host side to change any or all or these at the beginning of the question, c) as a potential down-the-road feature, provide a configuration dashboard that can be used to set/change these settings before a game is launched. 
5. On a related yet separate note: one feature idea/request that has emerged: as a newer teacher interested in incorporating error-based teaching methods, one may want to understand common mistakes/misconceptions from student perspectives: not only predefined distractors and explanations, but also common wrong answers and student examples that may be different from predefined distractors -- this might be something we keep in mind for Central

.
